src/spacyMatcher.py

```diff
@@ -1,54 +1,49 @@
-##    WORK_OF_ART - Titles of books, songs, etc.
-##    MONEY - Monetary values, including unit.
-##    QUANTITY - Measurements, as of weight or distance.
-##    ORDINAL - “first”, “second”, etc.
-##    CARDINAL - Numerals that do not fall under another type (not ordinal, quantity ..)
-##
-## It will return a list with distinct values of all occurrences of that entity in the text.
-###########################################################################################################
-
+"""
+This module contains a a function that returns named entities from text
+"""
 from aws_xray_sdk.core import xray_recorder
+import spacy
 
 @xray_recorder.capture('spacyMatcher')
-def spacyMatcher(text, tag): 
-    import spacy
-
-    if tag == '':
-        all_tags = ['PERSON','GPE','ORG','PERCENT','LANGUAGE','DATE','TIME','LOC','NORP','EVENT','WORK_OF_ART',
+def spacyMatcher(text, tag=None):
+    """
+    This function takes text and an optional tag.
+    It returns a dictionary tags and the entities corresponding to that tag.
+    
+    If no tag is specified, all tags will be used.
+    
+    Tags are SpaCy tags:
+    PERSON - People, including fictional.
+    ORG - Companies, agencies, institutions, etc.
+    GPE - Countries, cities, states.
+    PERCENT - Percentage, including ”%“.
+    LANGUAGE - Any named language.
+    DATE - Absolute or relative dates or periods.
+    TIME - Times smaller than a day.
+    LOC - Non-GPE locations, mountain ranges, bodies of water.
+    NORP - Nationalities or religious or political groups.
+    EVENT - Named hurricanes, battles, wars, sports events, etc.
+    WORK_OF_ART - Titles of books, songs, etc.
+    MONEY - Monetary values, including unit.
+    QUANTITY - Measurements, as of weight or distance.
+    ORDINAL - “first”, “second”, etc.
+    CARDINAL - Numerals that do not fall under another type (not ordinal, quantity ..).
+    """
+     all_tags = ['PERSON','GPE','ORG','PERCENT','LANGUAGE','DATE','TIME','LOC','NORP','EVENT','WORK_OF_ART',
                     'MONEY','QUANTITY','ORDINAL','CARDINAL']
-    else:
-        all_tags = [tag]
+        
+    tags = [tag] if tag else all_tags
 
-    all_topics = []
+    topics_by_tag = []
     if text != '-1':
         ### analysing the text 
         nlp = spacy.load("en_core_web_sm")
         doc = nlp(text)
-        interestWords = [(ent.text, ent.label_) for ent in doc.ents]
+        interesting_words = [(ent.text, ent.label_) for ent in doc.ents]
 
-        for tag in all_tags:
-            values_set = []
-            list_all_values = [tup[0] for tup in interestWords if any(i in tup for i in [tag])]        
-            values_set = set(list_all_values)
-            for obj in values_set:
-                all_topics.append({'type' : tag, 'topic' : obj})                    
+        for tag in tags:
+            values = {tup[0] for tup in interesting_words if any(i in tup for i in [tag])}        
```

Comment:
You can use a set comprehension since you want a set in the end. 

In general, it feels to me like this whole section could potentially be simplified even more but I'm not exactly clear on the details of how you need the final format to be.

Reply:
The object that we want to return should be a list of dictionaries, such as:
[{'type': 'PERSON', 'topic': 'Donald Trump'}, 
 {'type': 'GPE', 'topic': 'USA'}, 
 {'type': 'ORG', 'topic': 'BBC'}
]
I will definitely go through the code and see a better / simplified way to reach the same result.
Another great tip, many thanks!